File: python/day08.py
from collections import defaultdict
from operator import lt, gt, eq, ne, le, ge


with open('./inputs/08.txt') as f:
    instructions = f.readlines()

registers = defaultdict(int)
operators = {
    '<': lt,
    '>': gt,
    '==': eq,
    '!=': ne,
    '<=': le,
    '>=': ge,
}
maximum = 0

for line in instructions:
    register, instruction, amount, _, cond_reg, operator, value = line.split()
    directon = 1 if instruction == 'inc' else -1
    if operators[operator](registers[cond_reg], int(value)):
        registers[register] += directon * int(amount)
        maximum = max(registers[register], maximum)

print(max(registers.values()))
print(maximum)

File: nim/day08.nim
import tables, strutils

const instructions = readFile("./inputs/08.txt").splitLines()

type Instruction = tuple
  reg, instr, condReg, condOp: string
  amount, value: int


func parseLine(w: seq[string]): Instruction =
  (w[0], w[1], w[^3], w[^2], w[2].parseInt, w[^1].parseInt)

func isConditionSatisfied(op: string, regVal, value: int): bool =
  case op
    of "<":  regVal < value
    of ">":  regVal > value
    of "<=": regVal <= value
    of ">=": regVal >= value
    of "==": regVal == value
    else:    regVal != value


var
  registers = initCountTable[string]()
  maximum: int

for line in instructions:
  let
    l = line.split().parseLine()
    regVal = registers.getOrDefault(l.condReg)

  if isConditionSatisfied(l.condOp, regVal, l.value):
    case l.instr
      of "inc":
        registers.inc(l.reg, l.amount)
      of "dec":
        registers.inc(l.reg, -l.amount)
    maximum = max(registers[l.reg], maximum)


echo registers.largest.val
echo maximum
